.github/workflows/ci.yml

```diff
@@ -0,0 +1,141 @@
+# Builds the Betaflight firmware
+#
+# After building, artifacts are released to a separate repository.
+#
+# You'll also need to setup the follwing environment variables:
+#   repoNightly - The repository to release nightly builds to e.g. betaflight-configurator-nightly
+#   releaseNotes - The release notes to be published as part of the github release
+
+env:
+  repoNightly: betaflight/betaflight-pull-requests
+  releaseNotes: >
+    This is an automated development build.
+    It may be unstable and result in corrupted configurations or data loss.
+    **Use only for testing.**
+
+name: CI
+
+on:
+  push:
+    branches:
+      - master
+      - '*-maintenance'
+  pull_request:
+    branches:
+      - master
+      - '*-maintenance'
+
+jobs:
+  setup:
+    name: Setup
+    runs-on: ubuntu-20.04
+    outputs:
+      targets: ${{ steps.get-targets.outputs.targets }}
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Get all official build targets
+        id: get-targets
+        run: echo "::set-output name=targets::$(make targets-ci-print | jq -R -c 'split(" ")')"
+
+      - name: Cache build toolchain
+        uses: actions/cache@v2
+        id: cache-toolchain
+        with:
+          path: tools/
+          key: ${{ runner.os }}-${{ hashFiles('make/tools.mk') }}
+
+      - name: Download and install toolchain
+        if: steps.cache-toolchain.outputs.cache-hit != 'true'
+        run: make arm_sdk_install
+
+  build:
+    name: Build
+    needs: setup
+    runs-on: ubuntu-20.04
+    strategy:
+      matrix:
+        target: ${{ fromJson(needs.setup.outputs.targets) }}
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Fetch toolchain from cache
+        uses: actions/cache@v2
+        id: cache-toolchain
+        with:
+          path: tools/
+          key: ${{ runner.os }}-${{ hashFiles('make/tools.mk') }}
+
+      - name: Build target
+        run: make EXTRA_FLAGS=-Werror ${{ matrix.target }}
+
+      - name: Publish build artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          name: firmware
+          path: obj/*.hex
+
+  test:
+    name: Test
+    runs-on: ubuntu-20.04
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Install dependencies
+        run: sudo apt-get install -y libblocksruntime-dev
+
+      - name: Run sanity checks
+        run: make EXTRA_FLAGS=-Werror checks
+
+      - name: Run all unit tests
+        run: make EXTRA_FLAGS=-Werror test-all
```

Comment:
Shouldn't we test firstly then build?

Reply:
Running tasks in parallel is faster, for both failure and success.

Reply:
I think is cleaner to run tests first, and only if they pass, do the build. It will be some 5 minutes slower, but I remember when each build was about 45 minutes, not too much time ago :P

I don't know what others developers think about it.

Reply:
Since we run all builds in parallel here it doesn't change much to add another test job. I don't see a good reason to block the build while waiting for tests to complete. The extra resources needed to build while running a failing test job is negligible. If the tests pass but the build fails this is will fail faster and save time and resources. If successful resources used is basically the same and time is saved.